nvflare/fuel/f3/communicator.py

```diff
@@ -0,0 +1,216 @@
+
+    def send(self, endpoint: Endpoint, app_id: int, message: Message):
+        """Send a message to endpoint for app_id, no response is expected
+
+        Args:
+            endpoint: An endpoint to send the request to
+            app_id: Application ID
+            message: Message to send
+
+        Raises:
+            CommError: If any error happens while sending the data
+        """
+
+        self.conn_manager.send_message(endpoint, app_id, message.headers, message.payload)
+
+    def register_message_receiver(self, app_id: int, receiver: MessageReceiver):
+        """Register a receiver to process FCI message for the app
+
+        Args:
+            app_id: Application ID
+            receiver: The receiver to process the message
+
+        Raises:
+            CommError: If duplicate endpoint/app or receiver is of wrong type
+        """
+
+        self.conn_manager.register_message_receiver(app_id, receiver)
+
+    def add_connector(self, url: str, mode: Mode, secure: bool = False) -> str:
+        """Load a connector. The driver is selected based on the URL
+
+        Args:
+            url: The url to listen on or connect to, like "https://0:443". Use 0 for empty host
+            mode: Active for connecting, Passive for listening
+            secure: True if SSL is required.
+
+        Returns:
+            A handle that can be used to delete connector
+
+        Raises:
+            CommError: If any errors
+        """
+
+        return self._load_driver(url, mode, secure)
+
+    def start_listener(self, scheme: str, resources: dict) -> (str, str):
+        """Add and start a connector in passive mode on an address selected by the driver.
+
+        Args:
+            scheme: Connection scheme, e.g. http, https
+            resources: User specified resources like host and port ranges
+
+        Returns:
+            A tuple with connector handle and connect url
+
+        Raises:
+            CommError: If any errors like invalid host or port not available
+        """
+
+        driver_class = self.driver_mgr.find_driver_class(scheme)
+        if not driver_class:
+            raise CommError(CommError.NOT_SUPPORTED, f"No driver found for scheme {scheme}")
+
+        connect_url, listening_url = driver_class.get_urls(scheme, resources)
+        params = parse_url(listening_url)
+
+        handle = self.add_connector_advanced(driver_class(), Mode.PASSIVE, params, True)
+
+        return handle, connect_url
+
+    def add_connector_advanced(self, driver: Driver, mode: Mode, params: dict, start: bool = False) -> str:
+        """Add a connector using a specific driver instance.
+
+        Args:
+            driver: A transport driver instance
+            mode: Active or passive
+            params: Driver parameters
+            start: Start the connector if true
+
+        Returns:
+            A handle that can be used to delete the connector
+
+        Raises:
+            CommError: If any errors
+        """
+
```

Comment:
I think we should change the signature of this method to
```
def add_connector_advanced(self, driver: Driver, params: dict, mode: Mode, secure, start: bool = False) -> str:
```

And the code of `add_connector` should just utilize this method.

So should be:


```

    def add_connector(self, url: str, mode: Mode, secure: bool = False) -> str:
        """Adds a connector. The driver is selected based on the URL
        Args:
            url: The url to listen on or connect to, like "https://0:443". Use 0 for empty host
            mode: Active for connecting, Passive for listening
            secure: True if SSL is required.
        Returns:
            A handle that can be used to delete connector
        Raises:
            CommError: If any errors
        """
        driver_class = self.driver_mgr.find_driver_class(url)
        if not driver_class:
            raise CommError(CommError.NOT_SUPPORTED, f"No driver found for URL {url}")

        params = parse_url(url)
        return self.add_connector_advanced(driver=driver_class(), params=params, mode=mode, secure=secure, start=False)
        

```
And we don't need `_load_driver` method anymore.
If we really want that we should change it to

Reply:
Good suggestion. Removed 2 internal methods.